engine/access/ingestion/engine.go

```diff
@@ -172,30 +225,59 @@ func New(
 
 	defaultIndex, err := e.defaultProcessedIndex()
 	if err != nil {
-		return nil, fmt.Errorf("could not read default processed index: %w", err)
+		return nil, fmt.Errorf("could not read default finalized processed index: %w", err)
 	}
 
 	// create a jobqueue that will process new available finalized block. The `finalizedBlockNotifier` is used to
 	// signal new work, which is being triggered on the `processFinalizedBlockJob` handler.
 	e.finalizedBlockConsumer, err = jobqueue.NewComponentConsumer(
 		e.log.With().Str("module", "ingestion_block_consumer").Logger(),
 		e.finalizedBlockNotifier.Channel(),
-		processedHeight,
+		finalizedProcessedHeight,
 		finalizedBlockReader,
 		defaultIndex,
 		e.processFinalizedBlockJob,
-		workersCount,
+		processFinalizedBlocksWorkersCount,
 		searchAhead,
 	)
 	if err != nil {
 		return nil, fmt.Errorf("error creating finalizedBlock jobqueue: %w", err)
 	}
 
+	// jobqueue Jobs object that tracks sealed blocks by height. This is used by the txErrorMessagesConsumer
+	// to get a sequential list of sealed blocks.
+	sealedBlockReader := jobqueue.NewSealedBlockHeaderReader(state, headers)
+
+	// Create a job queue that will process error messages for new sealed blocks.
+	// It listens to block finalization events from `finalizedBlockNotifier`, then checks if there
+	// are new sealed blocks with `sealedBlockReader`. If there are, it starts workers to process
+	// them with `processTxResultErrorMessagesJob`, which fetches transaction error messages. At most
+	// `processTxErrorMessagesWorkersCount` workers will be created for concurrent processing.
+	// When a sealed block's error messages has been processed, it updates and persists the highest consecutive
+	// processed height with `txErrorMessagesProcessedHeight`. That way, if the node crashes,
+	// it reads the `txErrorMessagesProcessedHeight` and resume from `txErrorMessagesProcessedHeight + 1`.
+	// If the database is empty, rootHeight will be used to init the last processed height.
+	e.txErrorMessagesConsumer, err = jobqueue.NewComponentConsumer(
+		e.log.With().Str("module", "ingestion_tx_error_messages_consumer").Logger(),
+		e.finalizedBlockNotifier.Channel(),
```

Comment:
We should not reuse the same notifier channel. 

A notifier channel should only be used by only one consumer, if it's reused by multiple consumers, then only one of them will receive the notification, which means if a block is finalized, either `processFinalizedBlockJob` or `processTxResultErrorMessagesJob` will be called. But actually, we both of them to be called. 

To verify this behavior, we can probably add a test case where finalizing only one block should expect both collections and tx error messages are indexed. The test case would fail if the notification is only consumed once. 

I think we will need to create another notifier for `processTxResultErrorMessagesJob`. So that both consumers receive their notifications.

And the ingestion engine is getting quite complex now. Is there any particular reason this consumer has to be in the `ingestion/engine.go`? Would it make sense to be in a separate engine?

Reply:
Thank you, @zhangchiqing, for your comment! I have moved the core functionality to` tx_error_messages_core.go` and created a new, separate `engine` for the transaction error messages consumer to simplify the` ingestion/engine.go`. I have also added additional tests for this functionality. Could you please take a look at the updated version?